Duplicated payload definition and validation method for 2 methods with same payload

`goa gen` generates payload definition and validation method twice `create` and `update` methods using same payload.

## Example

```go
var _ = Service("user", func() {
	HTTP(func() {
		Path("/users")
	})

	Method("create", func() {
		Payload(UserPayload)
		Result(User)
		HTTP(func() {
			POST("")
			Body(UserPayload)
			Response(StatusCreated)
		})
	})

	Method("update", func() {
		Payload(func() {
			Attribute("id", Int)
			Attribute("user", UserPayload)
			Required("id", "user")
		})
		Result(User)
		HTTP(func() {
			PUT("/{id}")
			Body("user")
			Response(StatusOK)
		})
	})
})
```

`create` and `update` use `UserPayload`.

See https://github.com/akm/goa_v2_post_payload_including_snake_case_fields/blob/c22b6ae8e39937c3d118ac1cd61ca1190643caa6/design/desgin.go for more detail.

## Expected Behavior

Generate payload definition and validation method once only like this: https://github.com/akm/goa_v2_post_payload_including_snake_case_fields/blob/remove_duplicated_generated_code/gen/http/user/server/types.go

## Actual Behavior

Generate payload definition and validation method twice like this: https://github.com/akm/goa_v2_post_payload_including_snake_case_fields/blob/211fb042d3230eb20acf3d4fd9d55736156fac2d/gen/http/user/server/types.go

## Steps to Reproduce the Problem (including complete and simple design if relevant)

I made a project to reproduce this problem:
https://github.com/akm/goa_v2_post_payload_including_snake_case_fields

You can run the test by the following command:

```
$ git clone https://github.com/akm/goa_v2_post_payload_including_snake_case_fields.git $GOPATH/src/github.com/akm/goa_v2_post_payload_including_snake_case_fields
$ cd $GOPATH/src/github.com/akm/goa_v2_post_payload_including_snake_case_fields
$ go test
```

### current output

```
# github.com/akm/goa_v2_post_payload_including_snake_case_fields/gen/http/user/server
gen/http/user/server/types.go:26:6: UserPayload redeclared in this block
	previous declaration at gen/http/user/server/types.go:19:6
gen/http/user/server/types.go:100:50: ValidateUserPayload redeclared in this block
	previous declaration at gen/http/user/server/types.go:92:50
FAIL	github.com/akm/goa_v2_post_payload_including_snake_case_fields [build failed]
```

## Specifications

  - goa Version: v2.0.0-wip b70dea21c455aacc9404d1bb3e7861d2e1950029
  - Platform (Linux/Windows etc.): Mac OS X 10.14.2（18C54）

--- a/expr/http_body_types.go
+++ b/expr/http_body_types.go
@@ -12,16 +12,20 @@ import (
 // by removing the attributes of the method payload used to define headers and
 // parameters.
 func httpRequestBody(a *HTTPEndpointExpr) *AttributeExpr {
+	const suffix = "RequestBody"
+	var (
+		name = concat(a.Name(), "Request", "Body")
+	)
 	if a.Body != nil {
+		a.Body = DupAtt(a.Body)
+		renameType(a.Body, name, suffix)
 		return a.Body
 	}
 
-	const suffix = "RequestBody"
 	var (
 		payload   = a.MethodExpr.Payload
 		headers   = a.Headers
 		params    = a.Params
-		name      = concat(a.Name(), "Request", "Body")
 		userField string
 		passField string
 	)

--- a/expr/http_endpoint.go
+++ b/expr/http_endpoint.go
@@ -525,13 +525,9 @@ func (e *HTTPEndpointExpr) Finalize() {
 	init(e.Params)
 	init(e.Headers)
 
+	e.Body = httpRequestBody(e)
 	if e.Body != nil {
 		e.Body.Finalize()
-	}
-
-	if e.Body == nil {
-		// No explicit body, compute it
-		e.Body = httpRequestBody(e)
 	}
 
 	e.StreamingBody = httpStreamingBody(e)

--- a/http/codegen/client_body_types_test.go
+++ b/http/codegen/client_body_types_test.go
@@ -1,6 +1,7 @@
 package codegen
 
 import (
+	"bytes"
 	"testing"
 
 	"goa.design/goa/codegen"
@@ -53,6 +54,33 @@ func TestBodyTypeInit(t *testing.T) {
 			fs := clientType(genpkg, expr.Root.API.HTTP.Services[0], make(map[string]struct{}))
 			section := fs.SectionTemplates[c.SectionIndex]
 			code := codegen.SectionCode(t, section)
+			if code != c.Code {
+				t.Errorf("invalid code, got:\n%s\ngot vs. expected:\n%s", code, codegen.Diff(t, code, c.Code))
+			}
+		})
+	}
+}
+
+func TestClientTypes(t *testing.T) {
+	const genpkg = "gen"
+	cases := []struct {
+		Name string
+		DSL  func()
+		Code string
+	}{
+		{"multiple-methods", testdata.MultipleMethodsDSL, MultipleMethodsClientTypesFile},
+	}
+	for _, c := range cases {
+		t.Run(c.Name, func(t *testing.T) {
+			RunHTTPDSL(t, c.DSL)
+			fs := clientType(genpkg, expr.Root.API.HTTP.Services[0], make(map[string]struct{}))
+			var buf bytes.Buffer
+			for _, s := range fs.SectionTemplates[1:] {
+				if err := s.Write(&buf); err != nil {
+					t.Fatal(err)
+				}
+			}
+			code := codegen.FormatTestCode(t, "package foo\n"+buf.String())
 			if code != c.Code {
 				t.Errorf("invalid code, got:\n%s\ngot vs. expected:\n%s", code, codegen.Diff(t, code, c.Code))
 			}
@@ -151,5 +179,56 @@ func NewMethodExplicitBodyUserResultMultipleViewResulttypemultipleviewsOK(body *
 	}
 	res.C = c
 	return res
+}
+`
+
+const MultipleMethodsClientTypesFile = `// MethodARequestBody is the type of the "ServiceMultipleMethods" service
+// "MethodA" endpoint HTTP request body.
+type MethodARequestBody struct {
+	A *string ` + "`" + `form:"a,omitempty" json:"a,omitempty" xml:"a,omitempty"` + "`" + `
+}
+
+// MethodBRequestBody is the type of the "ServiceMultipleMethods" service
+// "MethodB" endpoint HTTP request body.
+type MethodBRequestBody struct {
+	A string               ` + "`" + `form:"a" json:"a" xml:"a"` + "`" + `
+	B *string              ` + "`" + `form:"b,omitempty" json:"b,omitempty" xml:"b,omitempty"` + "`" + `
+	C *APayloadRequestBody ` + "`" + `form:"c" json:"c" xml:"c"` + "`" + `
+}
+
+// APayloadRequestBody is used to define fields on request body types.
+type APayloadRequestBody struct {
+	A *string ` + "`" + `form:"a,omitempty" json:"a,omitempty" xml:"a,omitempty"` + "`" + `
+}
+
+// NewMethodARequestBody builds the HTTP request body from the payload of the
+// "MethodA" endpoint of the "ServiceMultipleMethods" service.
+func NewMethodARequestBody(p *servicemultiplemethods.APayload) *MethodARequestBody {
+	body := &MethodARequestBody{
+		A: p.A,
+	}
+	return body
+}
+
+// NewMethodBRequestBody builds the HTTP request body from the payload of the
+// "MethodB" endpoint of the "ServiceMultipleMethods" service.
+func NewMethodBRequestBody(p *servicemultiplemethods.PayloadType) *MethodBRequestBody {
+	body := &MethodBRequestBody{
+		A: p.A,
+		B: p.B,
+	}
+	if p.C != nil {
+		body.C = marshalServicemultiplemethodsAPayloadToAPayloadRequestBody(p.C)
+	}
+	return body
+}
+
+// ValidateAPayloadRequestBody runs the validations defined on
+// APayloadRequestBody
+func ValidateAPayloadRequestBody(body *APayloadRequestBody) (err error) {
+	if body.A != nil {
+		err = goa.MergeErrors(err, goa.ValidatePattern("body.a", *body.A, "patterna"))
+	}
+	return
 }
 `

--- a/http/codegen/server_types_test.go
+++ b/http/codegen/server_types_test.go
@@ -1,0 +1,116 @@
+package codegen
+
+import (
+	"bytes"
+	"testing"
+
+	"goa.design/goa/codegen"
+	"goa.design/goa/expr"
+	"goa.design/goa/http/codegen/testdata"
+)
+
+func TestServerTypes(t *testing.T) {
+	const genpkg = "gen"
+	cases := []struct {
+		Name string
+		DSL  func()
+		Code string
+	}{
+		{"multiple-methods", testdata.MultipleMethodsDSL, MultipleMethodsServerTypesFile},
+	}
+	for _, c := range cases {
+		t.Run(c.Name, func(t *testing.T) {
+			RunHTTPDSL(t, c.DSL)
+			fs := serverType(genpkg, expr.Root.API.HTTP.Services[0], make(map[string]struct{}))
+			var buf bytes.Buffer
+			for _, s := range fs.SectionTemplates[1:] {
+				if err := s.Write(&buf); err != nil {
+					t.Fatal(err)
+				}
+			}
+			code := codegen.FormatTestCode(t, "package foo\n"+buf.String())
+			if code != c.Code {
+				t.Errorf("invalid code, got:\n%s\ngot vs. expected:\n%s", code, codegen.Diff(t, code, c.Code))
+			}
+		})
+	}
+}
+
+const MultipleMethodsServerTypesFile = `// MethodARequestBody is the type of the "ServiceMultipleMethods" service
+// "MethodA" endpoint HTTP request body.
+type MethodARequestBody struct {
+	A *string ` + "`" + `form:"a,omitempty" json:"a,omitempty" xml:"a,omitempty"` + "`" + `
+}
+
+// MethodBRequestBody is the type of the "ServiceMultipleMethods" service
+// "MethodB" endpoint HTTP request body.
+type MethodBRequestBody struct {
+	A *string              ` + "`" + `form:"a,omitempty" json:"a,omitempty" xml:"a,omitempty"` + "`" + `
+	B *string              ` + "`" + `form:"b,omitempty" json:"b,omitempty" xml:"b,omitempty"` + "`" + `
+	C *APayloadRequestBody ` + "`" + `form:"c,omitempty" json:"c,omitempty" xml:"c,omitempty"` + "`" + `
+}
+
+// APayloadRequestBody is used to define fields on request body types.
+type APayloadRequestBody struct {
+	A *string ` + "`" + `form:"a,omitempty" json:"a,omitempty" xml:"a,omitempty"` + "`" + `
+}
+
+// NewMethodAAPayload builds a ServiceMultipleMethods service MethodA endpoint
+// payload.
+func NewMethodAAPayload(body *MethodARequestBody) *servicemultiplemethods.APayload {
+	v := &servicemultiplemethods.APayload{
+		A: body.A,
+	}
+	return v
+}
+
+// NewMethodBPayloadType builds a ServiceMultipleMethods service MethodB
+// endpoint payload.
+func NewMethodBPayloadType(body *MethodBRequestBody) *servicemultiplemethods.PayloadType {
+	v := &servicemultiplemethods.PayloadType{
+		A: *body.A,
+		B: body.B,
+	}
+	v.C = unmarshalAPayloadRequestBodyToServicemultiplemethodsAPayload(body.C)
+	return v
+}
+
+// ValidateMethodARequestBody runs the validations defined on MethodARequestBody
+func ValidateMethodARequestBody(body *MethodARequestBody) (err error) {
+	if body.A != nil {
+		err = goa.MergeErrors(err, goa.ValidatePattern("body.a", *body.A, "patterna"))
+	}
+	return
+}
+
+// ValidateMethodBRequestBody runs the validations defined on MethodBRequestBody
+func ValidateMethodBRequestBody(body *MethodBRequestBody) (err error) {
+	if body.A == nil {
+		err = goa.MergeErrors(err, goa.MissingFieldError("a", "body"))
+	}
+	if body.C == nil {
+		err = goa.MergeErrors(err, goa.MissingFieldError("c", "body"))
+	}
+	if body.A != nil {
+		err = goa.MergeErrors(err, goa.ValidatePattern("body.a", *body.A, "patterna"))
+	}
+	if body.B != nil {
+		err = goa.MergeErrors(err, goa.ValidatePattern("body.b", *body.B, "patternb"))
+	}
+	if body.C != nil {
+		if err2 := ValidateAPayloadRequestBody(body.C); err2 != nil {
+			err = goa.MergeErrors(err, err2)
+		}
+	}
+	return
+}
+
+// ValidateAPayloadRequestBody runs the validations defined on
+// APayloadRequestBody
+func ValidateAPayloadRequestBody(body *APayloadRequestBody) (err error) {
+	if body.A != nil {
+		err = goa.MergeErrors(err, goa.ValidatePattern("body.a", *body.A, "patterna"))
+	}
+	return
+}
+`

--- a/http/codegen/testdata/payload_decode_functions.go
+++ b/http/codegen/testdata/payload_decode_functions.go
@@ -3509,7 +3509,7 @@ var PayloadBodyNestedUserDecodeCode = `// DecodeMethodBodyUserRequest returns a 
 func DecodeMethodBodyUserRequest(mux goahttp.Muxer, decoder func(*http.Request) goahttp.Decoder) func(*http.Request) (interface{}, error) {
 	return func(r *http.Request) (interface{}, error) {
 		var (
-			body NestedType
+			body MethodBodyUserRequestBody
 			err  error
 		)
 		err = decoder(r).Decode(&body)
@@ -3519,7 +3519,7 @@ func DecodeMethodBodyUserRequest(mux goahttp.Muxer, decoder func(*http.Request) 
 			}
 			return nil, goa.DecodePayloadError(err.Error())
 		}
-		err = ValidateNestedType(&body)
+		err = ValidateMethodBodyUserRequestBody(&body)
 		if err != nil {
 			return nil, err
 		}

--- a/http/codegen/testdata/payload_dsls.go
+++ b/http/codegen/testdata/payload_dsls.go
@@ -2523,3 +2523,36 @@ var PayloadMultipartWithParamsAndHeadersDSL = func() {
 		})
 	})
 }
+
+var MultipleMethodsDSL = func() {
+	var APayload = Type("APayload", func() {
+		Attribute("a", String, func() {
+			Pattern("patterna")
+		})
+	})
+	var PayloadType = Type("PayloadType", func() {
+		Attribute("a", String, func() {
+			Pattern("patterna")
+		})
+		Attribute("b", String, func() {
+			Pattern("patternb")
+		})
+		Attribute("c", APayload)
+		Required("a", "c")
+	})
+	Service("ServiceMultipleMethods", func() {
+		Method("MethodA", func() {
+			Payload(APayload)
+			HTTP(func() {
+				POST("/")
+				Body(APayload)
+			})
+		})
+		Method("MethodB", func() {
+			Payload(PayloadType)
+			HTTP(func() {
+				PUT("/")
+			})
+		})
+	})
+}

